File: cpp/src/hangman.cc
#include <iostream>
#include <fstream>
#include <stdlib.h>
#include <algorithm>
#include <string>
#include <ctime>

using namespace std;

class Game
{
private:
  static const int maxFalseAttempts = 5;

  string getWord(const char* file);
  string renderResult(string word, string guesses);
public:
  void loop();
};

string Game::getWord(const char* file)
{
  ifstream wordFile(file);
  string word;
  int wordCount=0;
  while (getline(wordFile, word)) ++wordCount;
  int index = rand() % wordCount;
  wordFile.clear();
  wordFile.seekg(0, ios::beg);
  for (int n=0; getline(wordFile, word) && n<index; ++n);
  return word;
}

void Game::loop()
{
  const char* file = "dict_en.txt";
  string word = getWord(file);
  string guesses = "";
  int falseAttempts = 0;
  bool endOfGame = false;
  while(!endOfGame)
  {
    string result = renderResult(word, guesses);
    cout << result << endl;
    if(result == word)
    {
      endOfGame = true;
      cout << "You won!" << endl;
    }
    else
    {
      string guess;
      cout << "Guess a character: ";
      cin >> guess;
      if (guess.size() != 1)
      {
        cout << "Type a single character only! Try again..." << endl;
      }
      else if (!isalpha(guess[0]))
      {
        cout << "Type a alphabetic character! Try again..." << endl;
      }
      else if (guesses.find(tolower(guess[0])) != string::npos)
      {
        cout << "You already guessed this character! Try again..." << endl;
      }
      else
      {
        guesses += tolower(guess[0]);
        if (word.find(tolower(guess[0])) == string::npos)
        {
          ++falseAttempts;
          cout << "Wrong." << endl;
          if (1 == falseAttempts)
          {
            cout << "You have one false attempt." << endl;
          }
          else
          {
            cout << "You have " << falseAttempts << " false attempts." << endl;
          }
          if (falseAttempts > maxFalseAttempts)
          {
            endOfGame = true;
            cout << "You lost!" << endl;
            string upperWord = word;
            std::transform(upperWord.begin(), upperWord.end(),
              upperWord.begin(), ::toupper);
            cout << "The secret word has been " << upperWord << endl;
          }
        }
      }
    }
  }
}

string Game::renderResult(string word, string guesses)
{
  string result = "";
  for(unsigned int n=0; n<word.size(); ++n) result+= '_';
  for(unsigned int i=0; i<guesses.size(); ++i)
    for(unsigned int n=0; n<word.size(); ++n)
      if (guesses[i]==word[n]) result[n] = word[n];
  return result;
}

int main (int argc, const char * argv[])
{
  srand((unsigned)time(NULL));
  Game game;
  game.loop();
  return 0;
}

File: java/maven/src/main/java/hangman/Hangman.java
package hangman;

// Copyright (c) 2021, Oliver Merkel.
// Please see the AUTHORS file for details.
// All rights reserved.
//
// Use of this code is governed by a
// MIT license that can be found in the LICENSE file.

import java.io.File;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.nio.charset.Charset;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import java.util.Random;

public class Hangman {

  private class TestResult {
    String result;
    int incorrect;

    public TestResult( String word, String guesses ) {
      this.result = "";
      this.incorrect = 0;
      for (char c : word.toLowerCase().toCharArray()) {
        if (guesses.contains(Character.toString(c))) {
          result += c;
        }
        else {
          result += '_';
          this.incorrect++;
        }
      }
    }

    public String getResult() {
      return this.result;
    }

    public int getIncorrect() {
      return this.incorrect;
    }
  }

  private int maxFalseAttempts = 5;

  private void welcome() {
    System.out.println("Guess my secret word with a maximum of " +
      this.maxFalseAttempts + " false attempts.");
  }

  private Boolean playAgain() {
    List<String> yes = List.of( "y", "yes", "sure" );
    List<String> no = List.of( "n", "no", "nope" );
    Boolean isYes = false;
    Boolean isNo = false;
    while (!isYes && !isNo) {
      System.out.println("Play again? ");
      BufferedReader reader = new BufferedReader(
        new InputStreamReader(System.in));
      String answer = "";
      try { answer = reader.readLine().toLowerCase(); }
      catch(IOException ex) {
        System.out.println(ex.toString());
      }
      isYes = yes.contains(answer);
      isNo = no.contains(answer);
      if (!isYes && !isNo) {
        System.out.println("Sorry? Please answer with yes or no...");
      }
    }
    return isYes;
  }

  private String getWord() {
    String filename = "dict_en.txt";
    List<String> words = null;
    Random r = new Random();
    try{
      Path filePath = new File(filename).toPath();
      Charset charset = Charset.defaultCharset();
      words = Files.readAllLines(filePath, charset);
    }
    catch(IOException ex) {
      System.out.println(ex.toString());
      System.out.println("File not found:" + filename);
    }
    String word = "";
    while (word == "") {
      word = words.get(r.nextInt(words.size())).trim();
    }
    return word;
  }

  private void loop() {
    this.welcome();
    String word = this.getWord();
    Boolean endOfGame = false;
    int falseAttempts = 0;
    String guesses = "";
    while (!endOfGame) {
      TestResult test = (new TestResult(word, guesses));
      System.out.println(test.getResult());
      if (test.getIncorrect() == 0) {
        endOfGame = true;
        System.out.println("You won!");
      }
      else {
        System.out.println("Guess a character: ");
        BufferedReader reader = new BufferedReader(
          new InputStreamReader(System.in));
        String guess = "";
        try { guess = reader.readLine().toLowerCase(); }
        catch(IOException ex) {
          System.out.println(ex.toString());
        }
        if (guess.length() != 1) {
          System.out.println("Type a single character only! Try again...");
        }
        else if (!Character.isLetter(guess.charAt(0))) {
          System.out.println("Type a alphabetic character! Try again...");
        }
        else if (guesses.contains(guess)) {
          System.out.println("You already guessed this character! Try again...");
        }
        else {
          guesses += guess;
          if (!word.contains(guess)) {
            falseAttempts++;
            System.out.println("Wrong.");
            if (falseAttempts == 1) {
              System.out.println("You have one false attempt.");
            }
            else {
              System.out.println("You have " + falseAttempts + " false attempts.");
            }
            if (falseAttempts > maxFalseAttempts) {
              endOfGame = true;
              System.out.println("You lost!");
              System.out.println("The secret word has been " + word.toUpperCase());
            }
          }
        }
      }
    }
  }

  private void menuloop() {
    Boolean keepPlaying = true;
    while(keepPlaying) {
      this.loop();
      keepPlaying = this.playAgain();
    }
    System.out.println("Bye.");
  }

  public static void main(String[] args) {
    (new Hangman()).menuloop();
  }
}
